fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/translation/Translator.java

```diff
@@ -0,0 +1,226 @@
+import com.google.common.collect.Streams;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+
+/**
+ * This encapsulates a set of algorithms responsible for translating a given {@link Value} into another {@link Value} under
+ * a set of given assumptions.
+ * <br>
+ * for example, suppose we have the following {@link RecordConstructorValue}s:
+ * <ul>
+ *     <li>{@code P = RCV(a, b)}</li>
+ *     <li>{@code R = RCV(RCV(s, t), u, v)}</li>
+ * </ul>
+ * for a given {@code Value} such as {@code V = RCV(P.0, R.0.1)}, the translator can translate {@code V} in terms of <i>other</i>
+ * {@code Value}s, e.g. if we also have something like:
+ * <ul>
+ *     <li>{@code P` = RCV(RCV(c, a), RCV(p, b))}</li>
+ *     <li>{@code R` = RCV(w, x, RCV(RCV(t)))}</li>
+ * </ul>
+ * it can translate {@code V} into {@code RCV(P`.0.0, R`.2.0.0)}.
+ * <br>
+ * For certain shapes of {@code Value}s, multiple translations can be found, the {@code Translator} finds a translation
+ * that matches the maximum part of the {@code Value} with sub-{@code Value} on the other side. For example:
+ * <ul>
+ *     <li>{@code R = RCV(s+t, s, t)}</li>
+ * </ul>
+ * for a given {@code Value} such as {@code V = RCV(R.0)}, the translator can translate {@code V} in terms of <i>other</i>
+ * {@code Value}s, e.g. if we also have something like:
+ * <ul>
+ *     <li>{@code R` = RCV(s, t, (s+t))}</li>
+ * </ul>
+ * We could have the following translations for {@code V}:
+ * <ul>
+ *     <li>{@code RCV(R`.0 + R`.1}</li>
+ *     <li>{@code RCV(R`.0.0}</li>
+ * </ul>
+ * The translator is guaranteed to choose the second instead of the first one, i.e. it finds the maximum match possible.
+ * This property is very useful in the context of index matching where we want the query {@code Value} to be translated
+ * by means of the index {@code Value}, the query {@code Value} should reuse as many sub-{@code Value}s as possible from
+ * the index.
+ */
+public abstract class Translator {
+
+    @Nonnull
+    private final AliasMap aliasMap;
+
+    public Translator(@Nonnull AliasMap aliasMap) {
+        this.aliasMap = aliasMap;
+    }
+
+    /**
+     * Translates the {@link Value} {@code value} into another equivalent {@link Value} by means of the translation
+     * rules.
+     *
+     * @param value The {@code Value} to translate.
+     * @return an equivalent translated {@code Value}.
+     */
+    @Nonnull
+    public abstract Value translate(@Nonnull Value value);
+
+    /**
+     * Calculates the maximum sub-{@link Value}s in {@code rewrittenQueryValue} that has an exact match in the {@code candidateValue}.
+     * @param rewrittenQueryValue the query {@code Value}, it must be translated using {@code this} translator.
+     * @param candidateValue the candidate {@code Value} we want to search for maximum matches.
+     * @return A {@code Map} of all maximum matches.
+     */
+    @Nonnull
+    public MaxMatchMap calculateMaxMatches(@Nonnull final Value rewrittenQueryValue,
```

Comment:
hmm.. I think this is what the customer will end up using to generate an M3, right? (please see the tests in `ValueTranslationTest` for more information).

Are you hinting that a `Translator` should return another `Translator` instead? I think this might be even a better abstraction, albeit requiring a considerable amount of rework at this point.